
# Sales Dashboard EDA

This notebook contains SQL queries converted from the provided `.sql` file for performing exploratory data analysis (EDA) on the sales dashboard dataset. 
Each query is executed using the `ipython-sql` extension.

## Setup
Make sure you have the following prerequisites installed before running this notebook:
- `ipython-sql` for running SQL in the notebook.
- `psycopg2` for connecting to the PostgreSQL database.

### Steps to Connect
Replace `your_password` and `your_database` in the connection string below with your actual PostgreSQL credentials.

```python
%load_ext sql
%sql postgresql://postgres:your_password@localhost/your_database
```

---


In [20]:
import psycopg2
import pandas as pd

# Database connection
conn = psycopg2.connect(
    dbname="sales_dashboard",
    user="postgres",
    password="olajobi",
    host="localhost",  # Change if needed
    port="5432"        # Default PostgreSQL port
)

# Function to execute query and return as DataFrame
def execute_query(query):
    with conn.cursor() as cur:
        cur.execute(query)
        colnames = [desc[0] for desc in cur.description]
        rows = cur.fetchall()
    return pd.DataFrame(rows, columns=colnames)

### Query 1
```sql
--Rows count--
SELECT COUNT(*) AS total_rows FROM cleaned_sales_data
```

In [21]:
#Total Rows
query = "SELECT COUNT(*) AS total_rows FROM cleaned_sales_data;"
total_rows_df = execute_query(query)
print(total_rows_df)

   total_rows
0      536641


### Query 2
```sql
--Column statistics--
SELECT 
	COUNT(*) AS total_rows, 
	COUNT(DISTINCT InvoiceNo) AS unique_invoice,
	COUNT(DISTINCT CustomerID) AS unique_customers
 FROM cleaned_sales_data
```

In [22]:
#Column statistics
query = """
SELECT 
    COUNT(*) AS total_rows, 
    COUNT(DISTINCT InvoiceNo) AS unique_invoice,
    COUNT(DISTINCT CustomerID) AS unique_customers
FROM cleaned_sales_data;
"""
col_stats_df = execute_query(query)
print(col_stats_df)

   total_rows  unique_invoice  unique_customers
0      536641           25900              8082


We notice a discrepancy between the unique customers and unique invoices which suggests that multple purchases being done by the same customers

### Query 3
```sql
--Preview top rows--
SELECT * FROM cleaned_sales_data LIMIT 10
```

In [23]:
#Preview top rows
query = "SELECT * FROM cleaned_sales_data LIMIT 10;"
top_rows_df = execute_query(query)
print(top_rows_df)

  invoiceno stockcode                          description  quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   
5    536365     22752         SET 7 BABUSHKA NESTING BOXES         2   
6    536365     21730    GLASS STAR FROSTED T-LIGHT HOLDER         6   
7    536366     22633               HAND WARMER UNION JACK         6   
8    536366     22632            HAND WARMER RED POLKA DOT         6   
9    536367     84879        ASSORTED COLOUR BIRD ORNAMENT        32   

          invoicedate unitprice customerid         country  
0 2010-12-01 08:26:00      2.55    17850.0  United Kingdom  
1 2010-12-01 08:26:00      3.39    17850.0  United Kingdom  
2 2010-12-01 08:26:00   

### Query 4
```sql
--Sales Trend Over Time--
SELECT
	DATE_TRUNC('month', InvoiceDate) AS sales_month,
	SUM(Quantity * UnitPrice) AS total_sales
 FROM cleaned_sales_data
 GROUP BY sales_month
 ORDER BY sales_month
```

In [24]:
#Sales Trend Over Time
query = """
SELECT
    DATE_TRUNC('month', InvoiceDate) AS sales_month,
    SUM(Quantity * UnitPrice) AS total_sales
FROM cleaned_sales_data
GROUP BY sales_month
ORDER BY sales_month;
"""
sales_trend_df = execute_query(query)
print(sales_trend_df)


   sales_month  total_sales
0   2010-12-01    746723.61
1   2011-01-01    558448.56
2   2011-02-01    497026.41
3   2011-03-01    682013.98
4   2011-04-01   492367.841
5   2011-05-01    722094.10
6   2011-06-01    689977.23
7   2011-07-01   680156.991
8   2011-08-01    681386.46
9   2011-09-01  1017596.682
10  2011-10-01   1069368.23
11  2011-11-01   1456145.80
12  2011-12-01    432701.06


- The dataset shows a clear upward trend in total sales from December 2010 to November 2011.
- Notable peaks are observed in September, October, and November 2011, which coincide with the holiday season and possibly promotional activities. This indicates an increase in consumer spending and demand during these periods.
- January and February show a dip in total sales, possibly due to post-holiday spending cuts or the aftermath of promotional activities.
-  Retailers can prepare for increased demand during the peak months of September, October, and November.
-  For the reduced sales in January and February I suggest a need for sales promotions or targeted marketing to rejuvenate consumer interest post-holiday season (like new year diary and the sorts).

### Query 5
```sql
--Peak Sales Hours--
SELECT
	EXTRACT(hour FROM InvoiceDate) AS sales_hour,
	COUNT(*) AS transaction_count
 FROM cleaned_sales_data
 GROUP BY sales_hour
 ORDER BY transaction_count DESC
```

In [25]:
#Peak Sales Hours
query = """
SELECT
    EXTRACT(hour FROM InvoiceDate) AS sales_hour,
    COUNT(*) AS transaction_count
FROM cleaned_sales_data
GROUP BY sales_hour
ORDER BY transaction_count DESC;
"""
peak_hours_df = execute_query(query)
print(peak_hours_df)

   sales_hour  transaction_count
0          12              77573
1          15              76938
2          13              71247
3          14              66572
4          11              56949
5          16              54134
6          10              48808
7           9              34314
8          17              28371
9           8               8906
10         18               7941
11         19               3617
12         20                847
13          7                383
14          6                 41


Peak Hours:
- The highest transaction counts occur during the midday hours, specifically at 12 PM (noon) with 77,573 transactions, followed by 3 PM (76,938) and 1 PM (71,247).
This suggests that the majority of customers prefer shopping during their lunch breaks or early afternoons.

Active Periods:
- The busiest period is between 11 AM to 3 PM, where transaction counts consistently rank high.
This period may reflect a mix of consumer convenience (midday availability) and strategic sales timing by the business, possibly with promotions during this time.

Low Activity Hours:
- Early morning (6 AM to 9 AM) and evening hours (6 PM onwards) show significantly lower transaction counts.
The least active time is 6 AM, with only 41 transactions, indicating that few customers engage in shopping during early morning hours.

Implications for Staffing and Marketing:
- Businesses should allocate more staff and resources during peak hours, particularly from 11 AM to 3 PM, to manage the increased customer demand effectively.
Marketing campaigns and promotions could be scheduled for these peak hours to maximize visibility and impact.
Conversely, the business can reduce operational costs by minimizing staffing during non-peak hours (e.g., early morning or late evening).


### Query 6
```sql
--Top Customer by total sales--
SELECT
	CustomerID,
	SUM(Quantity * UnitPrice) AS total_spent
 FROM cleaned_sales_data
 GROUP BY CustomerID
 ORDER BY total_spent DESC
 LIMIT 10
```

In [26]:
#Top Customer by total sales
query = """
SELECT
    CustomerID,
    SUM(Quantity * UnitPrice) AS total_spent
FROM cleaned_sales_data
GROUP BY CustomerID
ORDER BY total_spent DESC
LIMIT 10;
"""
top_customers_df = execute_query(query)
print(top_customers_df)

  customerid total_spent
0    14646.0   279489.02
1    18102.0   256438.49
2    17450.0   187322.17
3    14911.0   132458.73
4    12415.0   123725.45
5    14156.0   113214.59
6    17511.0    88125.38
7    16684.0    65892.08
8    13694.0    62690.54
9    15311.0    59284.19


- The top three customers collectively spent 723,249.68, which is a significant portion of total sales revenue.
- Spending gradually decreases after the top few customers, with the 10th customer spending 59,284.19.
- The business could implement a tiered rewards program to encourage more spending from these high-value customers.

### Query 7
```sql
--Top Selling Products--
SELECT
	StockCode,
	Description,
	SUM(Quantity) AS total_quantity
 FROM cleaned_sales_data
 GROUP BY StockCode, Description
 ORDER BY total_quantity DESC
 LIMIT 10
```

In [12]:
#Top Selling Products
query = """
SELECT
    StockCode,
    Description,
    SUM(Quantity) AS total_quantity
FROM cleaned_sales_data
GROUP BY StockCode, Description
ORDER BY total_quantity DESC
LIMIT 10;
"""
top_selling_products_df = execute_query(query)
print(top_selling_products_df)

  stockcode                         description  total_quantity
0     84077   WORLD WAR 2 GLIDERS ASSTD DESIGNS           53751
1    85099B             JUMBO BAG RED RETROSPOT           47260
2     22197                      POPCORN HOLDER           36322
3     84879       ASSORTED COLOUR BIRD ORNAMENT           36282
4     21212     PACK OF 72 RETROSPOT CAKE CASES           36016
5    85123A  WHITE HANGING HEART T-LIGHT HOLDER           35006
6     23084                  RABBIT NIGHT LIGHT           30963
7     22492             MINI PAINT SET VINTAGE            26437
8     22616          PACK OF 12 LONDON TISSUES            26299
9     21977  PACK OF 60 PINK PAISLEY CAKE CASES           24719


- WORLD WAR 2 GLIDERS ASSTD DESIGNS (StockCode: 84077) is the top-selling product with 53,751 units sold, reflecting its broad appeal or frequent replenishment.
- JUMBO BAG RED RETROSPOT (StockCode: 85099B) ranks second with 47,260 units, followed by POPCORN HOLDER (StockCode: 22197) at 36,322 units.
- Prioritize availability for WORLD WAR 2 GLIDERS, JUMBO BAG RED RETROSPOT, and other top sellers to meet customer demand consistently.
- Highlight multipack items (e.g., PACK OF 72 RETROSPOT CAKE CASES) in promotions to attract bulk buyers.

### Query 8
```sql
--Top products by revenue--
SELECT
	StockCode,
	Description,
	SUM(Quantity * UnitPrice) AS total_revenue
 FROM cleaned_sales_data
 GROUP BY StockCode, Description
 ORDER BY total_revenue DESC
 LIMIT 10
```

In [13]:
#Top products by revenue
query = """
SELECT
    StockCode,
    Description,
    SUM(Quantity * UnitPrice) AS total_revenue
FROM cleaned_sales_data
GROUP BY StockCode, Description
ORDER BY total_revenue DESC
LIMIT 10;
"""
top_products_revenue_df = execute_query(query)
print(top_products_revenue_df)

  stockcode                         description total_revenue
0       DOT                      DOTCOM POSTAGE     206245.48
1     22423            REGENCY CAKESTAND 3 TIER     164459.49
2     47566                       PARTY BUNTING      98243.88
3    85123A  WHITE HANGING HEART T-LIGHT HOLDER      97659.94
4    85099B             JUMBO BAG RED RETROSPOT      92175.79
5     23084                  RABBIT NIGHT LIGHT      66661.63
6      POST                             POSTAGE      66230.64
7     22086     PAPER CHAIN KIT 50'S CHRISTMAS       63715.24
8     84879       ASSORTED COLOUR BIRD ORNAMENT      58792.42
9     79321                       CHILLI LIGHTS      53746.66


- The product with the highest revenue is DOTCOM POSTAGE (StockCode: DOT) with £206,245.48, indicating a significant volume of postage-related transactions.

- The REGENCY CAKESTAND 3 TIER (StockCode: 22423) ranks second, contributing £164,459.49, followed by PARTY BUNTING (StockCode: 47566) at £98,243.88.

- Other notable products like WHITE HANGING HEART T-LIGHT HOLDER (StockCode: 85123A) and JUMBO BAG RED RETROSPOT (StockCode: 85099B) showcase strong revenue contributions exceeding £90,000 each.

- Products like PAPER CHAIN KIT 50'S CHRISTMAS (StockCode: 22086) and RABBIT NIGHT LIGHT (StockCode: 23084) highlight the importance of seasonal and decorative items.

Insights for Action:
- Ensure sufficient stock and marketing for DOTCOM POSTAGE, REGENCY CAKESTAND 3 TIER, and other top products to sustain their revenue-generating potential.

- Seasonal items like PAPER CHAIN KIT 50'S CHRISTMAS and decorative products can be further promoted during peak holiday seasons to maximize sales.

### Query 9
```sql
--Top country by revenue--
 SELECT 
    Country,
    SUM(Quantity * UnitPrice) AS total_revenue
FROM cleaned_sales_data
GROUP BY Country
ORDER BY total_revenue DESC
LIMIT 10
```

In [31]:
#Top country by revenue
query = """
SELECT 
    Country,
    SUM(Quantity * UnitPrice) AS total_revenue
FROM cleaned_sales_data
GROUP BY Country
ORDER BY total_revenue DESC
LIMIT 10;
"""
top_countries_revenue_df = execute_query(query)
print(top_countries_revenue_df)

          country total_revenue
0  United Kingdom   8167128.184
1     Netherlands     284661.54
2            EIRE     262993.38
3         Germany     221509.47
4          France     197317.11
5       Australia     137009.77
6     Switzerland      56363.05
7           Spain      54756.03
8         Belgium      40910.96
9          Sweden      36585.41


United Kingdom's Significant Contribution:
- The United Kingdom generates an overwhelming majority of the revenue at approximately £8.17M, far surpassing all other countries. This underscores its position as the primary revenue driver for the business.

High Revenue from Netherlands and EIRE:
- The Netherlands (£284.66K) and EIRE (£262.99K) follow as the next largest contributors, but their revenue is significantly lower compared to the UK.

Leverage the UK Market:
- With the UK contributing the bulk of revenue, optimizing offerings for UK customers, such as premium products or loyalty programs, could yield even greater profitability.

Focus on Emerging Markets:
- Regions like Netherlands, EIRE, Germany, and France present opportunities for expansion. Strategic marketing and improved customer engagement in these markets could boost revenue.

### Query 10
```sql
--Number of transactions per country--
SELECT 
    Country,
    COUNT(DISTINCT InvoiceNo) AS transaction_count
FROM cleaned_sales_data
GROUP BY Country
ORDER BY transaction_count DESC
LIMIT 10
```

In [30]:
#Number of transactions per country
query = """
SELECT 
    Country,
    COUNT(DISTINCT InvoiceNo) AS transaction_count
FROM cleaned_sales_data
GROUP BY Country
ORDER BY transaction_count DESC
LIMIT 10;
"""
transactions_per_country_df = execute_query(query)
print(transactions_per_country_df)

          country  transaction_count
0  United Kingdom              23494
1         Germany                603
2          France                461
3            EIRE                360
4         Belgium                119
5           Spain                105
6     Netherlands                101
7     Switzerland                 74
8        Portugal                 71
9       Australia                 69


Dominance of the United Kingdom:
- The United Kingdom leads by a wide margin, accounting for 23,494 transactions, which is far more than any other country. This suggests it is the primary market for the business.

Secondary Markets:
- Countries like Germany (603), France (461), and EIRE (Ireland) (360) also contribute significantly but at a much lower scale compared to the UK.

Focus on Core Market:
- The UK's dominance indicates it is the key driver of business revenue. Efforts should be directed towards maintaining and growing this market through targeted marketing, inventory optimization, and efficient logistics.

Assessing Low-Performing Markets:
- Investigate why countries like Australia and Portugal have fewer transactions. It could be due to logistical challenges, lack of marketing efforts, or product relevance. Addressing these factors might increase market share.


### Query 11
```sql
--Identifying records with Incorrect UnitPrice--
SELECT * FROM cleaned_sales_data 
WHERE UnitPrice <= 0
```

In [16]:
#Identifying records with Incorrect UnitPrice
query = "SELECT * FROM cleaned_sales_data WHERE UnitPrice <= 0;"
incorrect_prices_df = execute_query(query)
print(incorrect_prices_df)

     invoiceno stockcode                          description  quantity  \
0       536414     22139     RETROSPOT TEA SET CERAMIC 11 PC         56   
1       536545     21134                       No description         1   
2       536546     22145      CHRISTMAS CRAFT HEART STOCKING          1   
3       536547     37509           NEW ENGLAND MUG W GIFT BOX         1   
4       536549    85226A                       No description         1   
...        ...       ...                                  ...       ...   
2507    580991     22481           Invalid Description Naming      -288   
2508    581190    84801A           Invalid Description Naming      -102   
2509    581234     72817  SET OF 2 CHRISTMAS DECOUPAGE CANDLE        27   
2510    581408     85175                CACTI T-LIGHT CANDLES        20   
2511    581422     23169           Invalid Description Naming      -235   

             invoicedate unitprice         customerid         country  
0    2010-12-01 11:52:00   

### Query 12
```sql
--Identifying invalid records
SELECT * FROM cleaned_sales_data WHERE quantity <= 0
```

In [17]:
#Identifying invalid records
query = """SELECT * FROM cleaned_sales_data WHERE quantity <= 0"""
invalid_records_df = execute_query(query)
print(invalid_records_df)

      invoiceno stockcode                        description  quantity  \
0       C536379         D         Invalid Description Naming        -1   
1       C536383    35004C    SET OF 3 COLOURED  FLYING DUCKS        -1   
2       C536391     22556     PLASTERS IN TIN CIRCUS PARADE        -12   
3       C536391     21984   PACK OF 12 PINK PAISLEY TISSUES        -24   
4       C536391     21983   PACK OF 12 BLUE PAISLEY TISSUES        -24   
...         ...       ...                                ...       ...   
10582   C581466     22838      3 TIER CAKE TIN RED AND CREAM        -1   
10583   C581466     22720  SET OF 3 CAKE TINS PANTRY DESIGN         -2   
10584   C581466     21216   SET 3 RETROSPOT TEA,COFFEE,SUGAR        -1   
10585   C581466     21535       RED RETROSPOT SMALL MILK JUG        -2   
10586   C581466     21232     STRAWBERRY CERAMIC TRINKET POT        -1   

              invoicedate unitprice customerid         country  
0     2010-12-01 09:41:00      27.5    14527.0

### Query 13
```sql
--Identifying repeated customers--
SELECT
    CustomerID,
    COUNT(DISTINCT InvoiceDate) AS num_purchases
FROM cleaned_sales_data
GROUP BY CustomerID
HAVING COUNT(DISTINCT InvoiceDate) > 1
ORDER BY num_purchases DESC
LIMIT 10
```

In [18]:
#Identifying repeated customers
query = """
SELECT
    CustomerID,
    COUNT(DISTINCT InvoiceDate) AS num_purchases
FROM cleaned_sales_data
GROUP BY CustomerID
HAVING COUNT(DISTINCT InvoiceDate) > 1
ORDER BY num_purchases DESC
LIMIT 10;
"""
repeated_customers_df = execute_query(query)
print(repeated_customers_df)

  customerid  num_purchases
0    14911.0            248
1    12748.0            225
2    17841.0            168
3    14606.0            129
4    15311.0            118
5    13089.0            114
6    12971.0             86
7    14527.0             85
8    14646.0             77
9    13408.0             77


High-Frequency Customers:
- CustomerID 14911.0 is the most frequent shopper, with 248 purchases, followed by CustomerID 12748.0 with 225 purchases.
- These top 10 customers represent a significant share of repeat business and are likely loyal customers

Strengthening Relationships:
- Implement a loyalty program to encourage continued engagement from these high-frequency customers.
- Consider offering exclusive discounts, early access to new products, or personalized services to incentivize repeat purchases.

### Query 14
```sql
--Total number of cancellations vs Total number of transactions
SELECT
    COUNT(*) FILTER (WHERE InvoiceNo LIKE 'C%') AS cancellations,
    COUNT(*) AS total_transactions
FROM cleaned_sales_data
```

In [19]:
#Total number of cancellations vs Total number of transactions
query = """
SELECT
    COUNT(*) FILTER (WHERE InvoiceNo LIKE 'C%') AS cancellations,
    COUNT(*) AS total_transactions
FROM cleaned_sales_data;
"""
cancellations_vs_transactions_df = execute_query(query)
print(cancellations_vs_transactions_df)

   cancellations  total_transactions
0           9251              536641


Cancellation Rate:
- Out of 536,641 total transactions, 9,251 were cancellations.
- This represents a cancellation rate of approximately 1.7% of all transactions, indicating that most transactions are successfully completed.

Operational Impact of Cancellations:
- Although cancellations constitute a small fraction of total transactions, their absolute number (9,251) is significant and could have tangible impacts on inventory management, logistics, and revenue stability.

Focus on Causes of Cancellations:
- Investigate the reasons behind these cancellations.

### Query 15
```sql
--Total number of unique original invoices vs unique cancelled invoice--
SELECT
    COUNT(DISTINCT InvoiceNo) AS original_invoices,
    COUNT(DISTINCT InvoiceNo) FILTER (WHERE InvoiceNo LIKE 'C%') AS cancelled_invoices 
FROM cleaned_sales_data
```

In [20]:
#Total number of unique original invoices vs unique cancelled invoice
query = """
SELECT
    COUNT(DISTINCT InvoiceNo) AS original_invoices,
    COUNT(DISTINCT InvoiceNo) FILTER (WHERE InvoiceNo LIKE 'C%') AS cancelled_invoices 
FROM cleaned_sales_data;
"""
original_vs_cancelled_df = execute_query(query)
print(original_vs_cancelled_df)

   original_invoices  cancelled_invoices
0              25900                3836


Proportion of Cancelled Invoices:
- Out of 25,900 total unique invoices, 3,836 are cancellations.
- This indicates that approximately 14.8% of all transactions are cancellations—a considerably higher cancellation rate.

Cancellation Trends:
- A cancellation rate close to 15% suggests a notable issue in the transaction process. It’s essential to identify whether this is due to customer behavior, operational inefficiencies, or product-related problems.

Insights for Action:
- High Cancellation Rate: Indicates a need for focused intervention. The specific periods, customer segments, or products contributing to these cancellations must be further analyzed.

### Query 16
```sql
--Number of cancellation each customer has made--
SELECT
    CustomerID,
    COUNT(InvoiceNo) AS num_cancellations
FROM cleaned_sales_data
WHERE InvoiceNo LIKE 'C%'
GROUP BY CustomerID
ORDER BY num_cancellations DESC
LIMIT 20
```

In [29]:
#Number of cancellation each customer has made
query = """
SELECT
    CustomerID,
    COUNT(InvoiceNo) AS num_cancellations
FROM cleaned_sales_data
WHERE InvoiceNo LIKE 'C%'
GROUP BY CustomerID
ORDER BY num_cancellations DESC
LIMIT 20;
"""
cancellations_by_customer_df = execute_query(query)
print(cancellations_by_customer_df)

   customerid  num_cancellations
0     14911.0                226
1     17841.0                136
2     17511.0                113
3     15311.0                112
4     12607.0                101
5     14410.0                 92
6     13798.0                 89
7     14606.0                 82
8     13113.0                 79
9     12471.0                 71
10    14680.0                 62
11    12415.0                 62
12    12472.0                 52
13    12709.0                 49
14    12748.0                 46
15    13534.0                 46
16    13069.0                 45
17    16801.0                 45
18    14426.0                 40
19    15005.0                 40


High-Cancellation Customers:
- The customer with ID 14911.0 has the highest number of cancellations (226), significantly more than the second-highest, 17841.0 (136 cancellations).
- The top 5 customers collectively account for a substantial number of cancellations, indicating that a small subset of customers is responsible for a large portion of the total cancellations.

Possible Issues:
- These customers might have experienced frequent issues, such as incorrect orders, defective products, or dissatisfaction with services.
- Alternatively, they might represent wholesale buyers or resellers with a higher volume of transactions, inherently leading to more cancellations.

Business Implications:
- These high-cancellation customers may require targeted engagement to understand and address their concerns.
- Monitoring the reason for cancellations for these specific customers could help uncover patterns and suggest corrective actions, such as improving product descriptions, shipping accuracy, or offering better customer service.

### Query 17
```sql
--Products that are most frequently cancelled--
SELECT
    StockCode,
    Description,
    COUNT(InvoiceNo) AS cancellations_count
FROM cleaned_sales_data
WHERE InvoiceNo LIKE 'C%'
GROUP BY StockCode, Description
ORDER BY cancellations_count DESC
LIMIT 10
```

In [28]:
#Products that are most frequently cancelled
query = """
SELECT
    StockCode,
    Description,
    COUNT(InvoiceNo) AS cancellations_count
FROM cleaned_sales_data
WHERE InvoiceNo LIKE 'C%'
GROUP BY StockCode, Description
ORDER BY cancellations_count DESC
LIMIT 10;
"""
cancellations_by_product_df = execute_query(query)
print(cancellations_by_product_df)

  stockcode                        description  cancellations_count
0         M         Invalid Description Naming                  244
1     22423           REGENCY CAKESTAND 3 TIER                  180
2      POST                            POSTAGE                  126
3     22960           JAM MAKING SET WITH JARS                   87
4         D         Invalid Description Naming                   77
5     22720  SET OF 3 CAKE TINS PANTRY DESIGN                    73
6         S                            SAMPLES                   60
7     21232     STRAWBERRY CERAMIC TRINKET BOX                   54
8     22699   ROSES REGENCY TEACUP AND SAUCER                    54
9     22666    RECIPE BOX PANTRY YELLOW DESIGN                   47


- The stock code "M" and "D" with "Invalid Description Naming" have the highest cancellation counts (244 and 77, respectively). These entries likely indicate data quality issues or misclassified products, which need further investigation.
- Among valid products, REGENCY CAKESTAND 3 TIER (StockCode: 22423) has the highest cancellations with 180, followed by POSTAGE (StockCode: POST) with 126.
- The business should investigate reasons for cancellations, focusing on the most frequently canceled products to identify and rectify underlying issues.
- Quality checks and customer feedback loops can help reduce cancellations by addressing customer concerns before they lead to order reversals.


### Query 18
```sql
--Analyzing cancellations with respect to time--
SELECT
    EXTRACT(month FROM InvoiceDate) AS month,
    COUNT(InvoiceNo) AS cancellations_count
FROM cleaned_sales_data
WHERE InvoiceNo LIKE 'C%'
GROUP BY month
ORDER BY cancellations_count DESC
```

In [23]:
#Analyzing cancellations with respect to time
query = """
SELECT
    EXTRACT(month FROM InvoiceDate) AS month,
    COUNT(InvoiceNo) AS cancellations_count
FROM cleaned_sales_data
WHERE InvoiceNo LIKE 'C%'
GROUP BY month
ORDER BY cancellations_count DESC;
"""
cancellations_by_time_df = execute_query(query)
print(cancellations_by_time_df)

   month  cancellations_count
0     10                 1194
1     12                 1078
2     11                 1073
3      9                  801
4      6                  709
5      1                  700
6      3                  696
7      7                  684
8      8                  667
9      5                  621
10     4                  558
11     2                  470


- October (Month 10) has the highest number of cancellations (1,194), followed by December (Month 12) with 1,078 cancellations, and November (Month 11) with 1,073 cancellations.
- These months align with major sales periods, such as holiday seasons or end-of-year promotions, where cancellations could be driven by buyer’s remorse, order errors, or out-of-stock products during high demand.
- Months with relatively fewer cancellations, such as February (470 cancellations) and April (558 cancellations), may correspond to low sales periods when customer transactions and cancellations are generally lower.
- Investigate specific reasons for cancellations during high-demand months like October, November, and December through customer feedback analysis or operational audits.